Mobs scatter all held items when gibbed

qwerltaz opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Gibbing now scatters the victim's inventory instead of perfectly putting it all in one spot.
- Also happens when the victim ashes due to temperature.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Pretty cool flavor. Makes the smite spell, minibombs, and trit fires have cooler effects.

## Technical details
<!-- Summary of code changes for easier review. -->
Add `TryUnequipAll` with arguments to scatter or throw items, use it in the gibbing method. 
- We return the unequipped entities. This is needed in, for example, the artifact crusher to insert all gibs in the crusher after gibbing (`ArtifactCrusherSystem.FinishCrushing`).


## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

Minibomb now scatters gibs and items everywhere

https://github.com/user-attachments/assets/a8dda4a2-8cf7-405c-a887-638df5942713


https://github.com/user-attachments/assets/48f92a66-e4ce-4b63-a7bb-7b95d852bf7b


https://github.com/user-attachments/assets/e1dcfa35-1fa2-4152-b190-7283159bd530

qwerltaz committed:
spread items of burned body on burning to ash

qwerltaz committed:
throw gibs and items on body gib

qwerltaz committed:
move to new method, add HasFilledDependentSlots

qwerltaz committed:
fix not throwing suit

qwerltaz committed:
return unequipped items

qwerltaz committed:
split dropping held items

qwerltaz committed:
oops whitespace

qwerltaz committed:
simplify docstring wording

qwerltaz committed:
revert old change

qwerltaz committed:
ensure gibs capacity at correct time

qwerltaz committed:
\n

qwerltaz committed:
dont enqueue already added item

qwerltaz committed:
add throw direction and cone support

qwerltaz committed:
split item throw into separate method

Princess-Cheeseballs reviewed (changes requested):
One thing I noticed.

slarticodefast reviewed (changes requested):
You can see this wildly mispredicting in the video. This is because the server and client are generating different random numbers for the item location. Gibbing has the same bug at the moment.
There is a workaround by setting the seed for the RNG, but ideally this would be using RandomPredicted https://github.com/space-wizards/RobustToolbox/pull/5849

qwerltaz committed:
move throwing to its system

qwerltaz committed:
'predict' item scattering

Disabled scattering for current usages of TryThrowRandom because the effects look better that way imo, so the prediction improvemed will be less visible

qwerltaz committed:
rename TryThrowRandom

qwerltaz committed:
reword docstring for TryUnequipAllAndScatter

Princess-Cheeseballs commented on Content.Shared/Throwing/ThrowingSystem.cs:
```diff
@@ -245,4 +249,56 @@ public void TryThrow(EntityUid uid,
+        var maxThrowImpulse = maxThrowImpulseModifier * UnequipAllDefaultMaxThrowForce;
+
+        // TODO: Replace with RandomPredicted once the engine PR is merged (#5849)
+        var rand = new System.Random((int)_gameTiming.CurTick.Value);
```
Use the predicted randomness workaround that's used in SharedEntityEffects, and SharedBloodstream

qwerltaz committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into feat/more-item-dropping

qwerltaz committed:
remove hasHands bool, always drop item

qwerltaz committed:
new random seed for each thrown item

qwerltaz committed:
rename UnequipAllDefaultMaxThrowForce from inventory-oriented name to ThrowManyRandomDefaultMaxForce (i forgor)

qwerltaz committed:
make items scatter in a circle, not square

qwerltaz committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into feat/more-item-dropping

# Conflicts:
#	Content.Shared/Inventory/InventorySystem.Equip.cs

Princess-Cheeseballs commented on Content.Shared/Throwing/ThrowingSystem.cs:
```diff
@@ -243,4 +247,62 @@ public void TryThrow(EntityUid uid,
+            }
+
+            var throwSpeed = _random.NextFloat() * maxThrowImpulse;
+            TryThrow(item, currentDir * throwSpeed, throwSpeed * 2.0f + 2.0f);
```
Magic numbers seem weird here especially considering we already have several modifiers available. I'd just remove them and pass in throwSpeed unadjusted or not at all.

Princess-Cheeseballs reviewed (changes requested):
Code looks good just one change.

qwerltaz committed:
less throwSpeed variance, add air time variance

qwerltaz committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into feat/more-item-dropping

# Conflicts:
#	Content.Shared/Throwing/ThrowingSystem.cs

qwerltaz committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into feat/more-item-dropping

qwerltaz committed:
remove magic number from TryThrowManyRandom

Princess-Cheeseballs commented on Content.Shared/Throwing/ThrowingSystem.cs:
```diff
@@ -255,7 +259,64 @@ public void TryThrow(EntityUid uid,
+            }
 
+            var throwSpeed = _random.NextFloat(0.2f, 1.0f) * maxThrowImpulse;
+            var airTimeVariance = _random.NextFloat(1.0f, 2.0f);
```
Include the magic numbers as a const somewhere, otherwise looks fine.

Princess-Cheeseballs reviewed (changes requested):
One change and then I can approve.

qwerltaz committed:
constants for throw speed and air time variance

Princess-Cheeseballs reviewed (approved):
Awesome. Needs one more approval still.

qwerltaz committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into feat/more-item-dropping

# Conflicts:
#	Content.Shared/Throwing/ThrowingSystem.cs

qwerltaz committed:
Merge remote-tracking branch 'upstream/master' into feat/more-item-dropping

# Conflicts:
#	Content.Server/Destructible/Thresholds/Behaviors/BurnBodyBehavior.cs

qwerltaz committed:
Merge remote-tracking branch 'upstream/master' into feat/more-item-dropping

# Conflicts:
#	Content.Shared/Body/Systems/SharedBodySystem.Body.cs
#	Content.Shared/Magic/SharedMagicSystem.cs

qwerltaz committed:
gibbing system: use TryThrowManyRandom, inventory system: subscribe to being gibbed

qwerltaz committed:
support scatter giblets on gib, add to gib behaviour

gib behaviour is triggered on explosions. scattering the giblets makes them fly in all directions instead of all perfectly flying and landing in the same quentum superposition. for other gib sources like the smite spellbook, this is not an issue so scatterGiblets = false by default.

qwerltaz committed:
rename TryUnequipAllAndScatter to TryUnequipAllAndThrow

qwerltaz committed:
smaller giblet launch impulse on body destroyed

this removes giblets flying far even when e.g., gibbed by normal blunt weapon, but still makes giblets fly realisticallly in all directions when gibbed by minibomb (because scatterGiblets is true)

qwerltaz committed:
TryThrowManyRandom: scatter after throwing

qwerltaz commented:
Yes, with some tweaks to make the gibbing the same as in my original version. done.

qwerltaz committed:
Merge remote-tracking branch 'upstream/master' into feat/more-item-dropping

# Conflicts:
#	Content.Shared/Gibbing/GibbingSystem.cs
#	Content.Shared/Inventory/InventorySystem.Equip.cs
#	Content.Shared/Throwing/ThrowingSystem.cs

qwerltaz committed:
oops

qwerltaz committed:
smite spell throws giblets with FORCE
